File: spec/java/src/io/kaitai/struct/specwrite/TestBytesEosPadTerm.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ByteBufferKaitaiStream;
import io.kaitai.struct.ConsistencyError;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.BytesEosPadTerm;
import org.testng.annotations.Test;

public class TestBytesEosPadTerm extends CommonSpec {
    @Test
    public void testCheckGoodMaxLengths() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "12345678901234567890".getBytes());
        setStrTermInclude(r, "12345678901234567890".getBytes());
        check(r);
    }

    @Test
    public void testCheckGoodMinLengths() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "".getBytes());
        setStrTerm(r, "".getBytes());
        setStrTermAndPad(r, "".getBytes());
        setStrTermInclude(r, "@".getBytes());
        check(r);
    }

    @Test(expectedExceptions = java.nio.BufferOverflowException.class)
    public void testCheckLongerStrPad() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "123456789012345678901".getBytes());
        check(r);
    }

    @Test
    public void testCheckGoodLastByteStrPad() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, ("@@@@@"+"@@@@@"+"@@@@@"+"@@@@?").getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "12345678901234567890".getBytes());
        setStrTermInclude(r, "12345678901234567890".getBytes());
        check(r);
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_pad,.*")
    public void testCheckBadLastByteStrPad() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "123456789012345678@".getBytes());
        check(r);
    }

    @Test(expectedExceptions = java.nio.BufferOverflowException.class)
    public void testCheckLongerStrTerm() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "123456789012345678901".getBytes());
        check(r);
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term,.*")
    public void testCheckBadHasTerminator1StrTerm() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "123456789012@4567890".getBytes());
        check(r);
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term,.*")
    public void testCheckBadHasTerminator2StrTerm() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "1234567890123456789@".getBytes());
        check(r);
    }

    @Test(expectedExceptions = java.nio.BufferOverflowException.class)
    public void testCheckLongerStrTermAndPad() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "123456789012345678901".getBytes());
        check(r);
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_and_pad,.*")
    public void testCheckBadHasTerminatorStrTermAndPad() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "@2345678901234567890".getBytes());
        check(r);
    }

    @Test
    public void testCheckGoodLastByte1StrTermAndPad() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, ("+++++"+"+++++"+"+++++"+"++++").getBytes());
        setStrTermInclude(r, "12345678901234567890".getBytes());
        check(r);
    }

    @Test
    public void testCheckGoodLastByte2StrTermAndPad() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, ("+++++"+"+++++"+"+++++"+"++++0").getBytes());
        setStrTermInclude(r, "12345678901234567890".getBytes());
        check(r);
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_and_pad,.*")
    public void testCheckBadLastByteStrTermAndPad() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "1234567890123456789+".getBytes());
        check(r);
    }

    @Test(expectedExceptions = java.nio.BufferOverflowException.class)
    public void testCheckLongerStrTermInclude() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "12345678901234567890".getBytes());
        setStrTermInclude(r, "123456789012345678901".getBytes());
        check(r);
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_include,.*")
    public void testCheckEmptyStrTermInclude() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "12345678901234567890".getBytes());
        setStrTermInclude(r, "".getBytes());
        check(r);
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_include,.*")
    public void testCheckBadNoTerminatorStrTermInclude() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "12345678901234567890".getBytes());
        setStrTermInclude(r, "123".getBytes());
        check(r);
    }

    @Test
    public void testCheckGoodTerminator1StrTermInclude() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "12345678901234567890".getBytes());
        setStrTermInclude(r, "12@".getBytes());
        check(r);
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_include,.*")
    public void testCheckEarlyTerminator1StrTermInclude() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "12345678901234567890".getBytes());
        setStrTermInclude(r, "1@@".getBytes());
        check(r);
    }

    @Test
    public void testCheckGoodTerminator2StrTermInclude() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "12345678901234567890".getBytes());
        setStrTermInclude(r, "1234567890123456789@".getBytes());
        check(r);
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_include,.*")
    public void testCheckEarlyTerminator2StrTermInclude() throws Exception {
        BytesEosPadTerm r = new BytesEosPadTerm();
        setStrPad(r, "12345678901234567890".getBytes());
        setStrTerm(r, "12345678901234567890".getBytes());
        setStrTermAndPad(r, "12345678901234567890".getBytes());
        setStrTermInclude(r, "123456789012345678@@".getBytes());
        check(r);
    }

    private static void setStrPad(BytesEosPadTerm r, byte[] value) {
        BytesEosPadTerm.StrPadType s = new BytesEosPadTerm.StrPadType(null, r, r._root());
        s.setValue(value);
        s._check();

        r.setStrPad(s);
    }

    private static void setStrTerm(BytesEosPadTerm r, byte[] value) {
        BytesEosPadTerm.StrTermType s = new BytesEosPadTerm.StrTermType(null, r, r._root());
        s.setValue(value);
        s._check();

        r.setStrTerm(s);
    }

    private static void setStrTermAndPad(BytesEosPadTerm r, byte[] value) {
        BytesEosPadTerm.StrTermAndPadType s = new BytesEosPadTerm.StrTermAndPadType(null, r, r._root());
        s.setValue(value);
        s._check();

        r.setStrTermAndPad(s);
    }

    private static void setStrTermInclude(BytesEosPadTerm r, byte[] value) {
        BytesEosPadTerm.StrTermIncludeType s = new BytesEosPadTerm.StrTermIncludeType(null, r, r._root());
        s.setValue(value);
        s._check();

        r.setStrTermInclude(s);
    }

    private static void check(BytesEosPadTerm r) {
        final byte[] buf = "12345678901234567890".getBytes();
        if (r.strPad() == null) {
            setStrPad(r, buf);
        }
        if (r.strTerm() == null) {
            setStrTerm(r, buf);
        }
        if (r.strTermAndPad() == null) {
            setStrTermAndPad(r, buf);
        }
        if (r.strTermInclude() == null) {
            setStrTermInclude(r, buf);
        }
        r._check();
        r._write(new ByteBufferKaitaiStream(80));
    }

    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return BytesEosPadTerm.class;
    }

    @Override
    protected String getSrcFilename() {
        return "str_pad_term.bin";
    }
}

File: spec/python/specwrite/test_bytes_eos_pad_term.py
import io
import unittest
from kaitaistruct import ConsistencyError, KaitaiStream
from specwrite.common_spec import CommonSpec

from testwrite.bytes_eos_pad_term import BytesEosPadTerm

class TestBytesEosPadTerm(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super(TestBytesEosPadTerm, self).__init__(*args, **kwargs)
        self.struct_class = BytesEosPadTerm
        self.src_filename = 'src/str_pad_term.bin'

    def test_check_good_max_lengths(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"12345678901234567890")
        set_str_term_include(r, b"12345678901234567890")
        check(r)

    def test_check_good_min_lengths(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"")
        set_str_term(r, b"")
        set_str_term_and_pad(r, b"")
        set_str_term_include(r, b"@")
        check(r)

    def test_check_longer_str_pad(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"123456789012345678901")
        with self.assertRaisesRegex(EOFError, u"^requested to write \\d+ bytes, but only \\d+ bytes left in the stream$"):
            check(r)

    def test_check_good_last_byte_str_pad(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"@@@@@" b"@@@@@" b"@@@@@" b"@@@@?")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"12345678901234567890")
        set_str_term_include(r, b"12345678901234567890")
        check(r)

    def test_check_bad_last_byte_str_pad(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"123456789012345678@")
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_pad,"):
            check(r)

    def test_check_longer_str_term(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"123456789012345678901")
        with self.assertRaisesRegex(EOFError, u"^requested to write \\d+ bytes, but only \\d+ bytes left in the stream$"):
            check(r)

    def test_check_bad_has_terminator1_str_term(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"123456789012@4567890")
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term,"):
            check(r)

    def test_check_bad_has_terminator2_str_term(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"1234567890123456789@")
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term,"):
            check(r)

    def test_check_longer_str_term_and_pad(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"123456789012345678901")
        with self.assertRaisesRegex(EOFError, u"^requested to write \\d+ bytes, but only \\d+ bytes left in the stream$"):
            check(r)

    def test_check_bad_has_terminator_str_term_and_pad(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"@2345678901234567890")
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_and_pad,"):
            check(r)

    def test_check_good_last_byte1_str_term_and_pad(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"+++++" b"+++++" b"+++++" b"++++")
        set_str_term_include(r, b"12345678901234567890")
        check(r)

    def test_check_good_last_byte2_str_term_and_pad(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"+++++" b"+++++" b"+++++" b"++++0")
        set_str_term_include(r, b"12345678901234567890")
        check(r)

    def test_check_bad_last_byte_str_term_and_pad(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"1234567890123456789+")
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_and_pad,"):
            check(r)

    def test_check_longer_str_term_include(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"12345678901234567890")
        set_str_term_include(r, b"123456789012345678901")
        with self.assertRaisesRegex(EOFError, u"^requested to write \\d+ bytes, but only \\d+ bytes left in the stream$"):
            check(r)

    def test_check_empty_str_term_include(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"12345678901234567890")
        set_str_term_include(r, b"")
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_include,"):
            check(r)

    def test_check_bad_no_terminator_str_term_include(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"12345678901234567890")
        set_str_term_include(r, b"123")
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_include,"):
            check(r)

    def test_check_good_terminator1_str_term_include(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"12345678901234567890")
        set_str_term_include(r, b"12@")
        check(r)

    def test_check_early_terminator1_str_term_include(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"12345678901234567890")
        set_str_term_include(r, b"1@@")
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_include,"):
            check(r)

    def test_check_good_terminator2_str_term_include(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"12345678901234567890")
        set_str_term_include(r, b"1234567890123456789@")
        check(r)

    def test_check_early_terminator2_str_term_include(self):
        r = BytesEosPadTerm()
        set_str_pad(r, b"12345678901234567890")
        set_str_term(r, b"12345678901234567890")
        set_str_term_and_pad(r, b"12345678901234567890")
        set_str_term_include(r, b"123456789012345678@@")
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_include,"):
            check(r)


def set_str_pad(r, value):
    s = BytesEosPadTerm.StrPadType(None, r, r._root)
    s.value = value
    s._check()

    r.str_pad = s


def set_str_term(r, value):
    s = BytesEosPadTerm.StrTermType(None, r, r._root)
    s.value = value
    s._check()

    r.str_term = s


def set_str_term_and_pad(r, value):
    s = BytesEosPadTerm.StrTermAndPadType(None, r, r._root)
    s.value = value
    s._check()

    r.str_term_and_pad = s


def set_str_term_include(r, value):
    s = BytesEosPadTerm.StrTermIncludeType(None, r, r._root)
    s.value = value
    s._check()

    r.str_term_include = s


def check(r):
    buf = b"12345678901234567890"
    if not hasattr(r, 'str_pad'):
        set_str_pad(r, buf)
    if not hasattr(r, 'str_term'):
        set_str_term(r, buf)
    if not hasattr(r, 'str_term_and_pad'):
        set_str_term_and_pad(r, buf)
    if not hasattr(r, 'str_term_include'):
        set_str_term_include(r, buf)

    r._check()
    ks_io = KaitaiStream(io.BytesIO(bytearray(80)))
    r._write(ks_io)
